### DISCUSSION TITLE
Apple receipt validation

### DISCUSSION THREAD

**Discussion by sabina1997** [imdhemy/laravel-in-app-purchases#380] [Q&A] [answered]
Hi there,
Im trying to implement liap , I have created the event listeners and was able to test in sandbox mode. Everything works good.
When submitting the app in app store, I got the following message:

_When validating receipts on your server, your server needs to be able to handle a production-signed app getting its receipts from Apple’s test environment...._

So It looks like I missed the receipt validation. I wanted to know where to validate the receipts, can I handle it on event listeners or it needs to be separated and what is the flow? Im handling everything through events.
Thanks

**Comment by imdhemy** [comment, marked as answer]
Hi @sabina1997,

For information on receipt verification, please refer to the [documentation](https://imdhemy.com/laravel-iap-docs/). The documentation covers:

- [Verification of product receipts](https://imdhemy.com/laravel-iap-docs/docs/app-store/sell-products#validate-receipts).
- [Verification of subscription receipts](https://imdhemy.com/laravel-iap-docs/docs/app-store/sell-subscriptions).

Ensure that the receipt is transmitted from the mobile application to the backend, a process that can be implemented in a controller.

> [!NOTE]  
> While this flow is deprecated, it still functions effectively. I am available to assist you with either this approach or the newly introduced one by Apple. This assistance aids in maintaining the LIAP package, especially as I currently lack access to an iOS App. Let me know if this solution meets your needs. Feel free to reach out to me on [LinkedIn](https://www.linkedin.com/in/imdhemy/) or via [email](https://github.com/imdhemy/laravel-in-app-purchases/blob/1.x/composer.json#L16) for further discussion.

**Reply by sabina1997** [reply]
Thank you @imdhemy , I was able to implement apple successfully, now I have problems with google authentication but will ask you via email maybe. Thanks for your support.
